File: dart/phone-number/lib/phone_number.dart
class PhoneNumber {
  // invalid if area code or exchange starts with 0 or 1
  bool isValid(String s) => (
    s[0] != '0' && s[0] != '1' && s[3] != '0' && s[3] != '1'
  );

  String clean(String number) {
    var cleaned = number.replaceAll(new RegExp(r'[^\d]'), '');
    if (cleaned.length != 10) {
      if (cleaned.length == 11) {
        if (cleaned[0] == '1') {
          cleaned = cleaned.substring(1);
          if (isValid(cleaned)) {
            return cleaned;
          }
        }
      }
      return null;
    }
    return isValid(cleaned) ? cleaned : null;
  }
}

File: csharp/phone-number/PhoneNumber.cs
using System;
using System.Text.RegularExpressions;

public class PhoneNumber
{
    public static string Clean(string phoneNumber)
    {
        Regex re = new Regex("[^0-9]");
        string cleanedNumber = re.Replace(phoneNumber, "");
        if (cleanedNumber.Length < 10 || cleanedNumber.Length > 11) {
            throw new ArgumentException("Invalid phone number length.");
        }
        if (cleanedNumber.Length == 11) {
            if (cleanedNumber[0] != '1') {
                throw new ArgumentException("Invalid 11-digit phone number.");
            } else {
                cleanedNumber = cleanedNumber.Substring(1);
            }
        }
        if (cleanedNumber[0] == '0' || cleanedNumber[0] == '1' || cleanedNumber[3] == '0' || cleanedNumber[3] == '1') {
            throw new ArgumentException("Invalid area code or exchange.");
        }
        return cleanedNumber;
    }
}